src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentState = {};
+    }
+
+    if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    }
+
+    if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
```

Comment:
Cloning `currentState` here is correct before modifying it. However, similar to the initial cloning, if `currentState` contains nested objects, consider using a deep clone to prevent mutations of nested objects.